### PR TITLE
Introduce support for health rules

### PR DESCRIPTION
**Pull request by idoru**
Supports AlwaysHealthy and SingleConditionTypes

When no health rule is specified, Healthy is Unknown until valid outputs are produced from a resources, at which time they will Healthy will become True.

Updates #812

Follows specification from https://github.com/vmware-tanzu/cartographer/pull/738 with the following exceptions:

When no health rule is specified for a template, the reason for a resource's Healthy condition may also be attributed these other values, not originally in the RFC:

- NoResource
- OutputsAvailable
- OutputNotAvailable

Parts not implemented in this PR:

- MultiMatch rules
- Healthy condition in owner conditions

## Release Note

Support reporting Healthy status on resources statuses.

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [x] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [x] Removed non-atomic or `wip` commits
- [x] Filled in the [Release Note](#Release-Note) section above 
- [x] Modified the docs to match changes

### CODE REVIEW

File: pkg/realizer/healthcheck/healthcheck_test.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ import (
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 	. "github.com/onsi/gomega/gstruct"
+	v1 "k8s.io/api/core/v1"
```
Comment (cirocosta) [inline pkg/realizer/healthcheck/healthcheck_test.go]: we usually name the import as `corev1` 😁 

```suggestion
	corev1 "k8s.io/api/core/v1"
```

File: pkg/realizer/healthcheck/healthcheck.go [hunk 2]
```diff
@@ -31,31 +31,33 @@ func DetermineHealthCondition(rule *v1alpha1.HealthRule, realizedResource *v1alp
...
+			return conditions.AlwaysHealthyResourcesHealthyCondition()
 		}
-		if resultString, ok := result[0].(string); ok {
-			conditionStatus := metav1.ConditionStatus(resultString)
-			if conditionStatus == metav1.ConditionFalse || conditionStatus == metav1.ConditionTrue {
-				return conditions.SingleConditionMatchCondition(conditionStatus, rule.SingleConditionType)
-			} else {
-				return conditions.SingleConditionMatchCondition(metav1.ConditionUnknown, rule.SingleConditionType)
+		if rule.SingleConditionType != "" && stampedObject != nil {
+			jsonpathQuery := fmt.Sprintf("{.status.conditions[?(@.type==\"%s\")].status}", rule.SingleConditionType)
```
Comment (cirocosta) [inline pkg/realizer/healthcheck/healthcheck.go:54]: another option here would be to do something like

```go
type Conditionable struct {
        Status struct {
                Conditions []metav1.Condition `json:"conditions,omitempty"`
        } `json:"status,omitempty"`
}
```

which could be populated by unmarshalling into it; `Conditonable` could then have some helper methods to do what we're doing here

```go
func (c Conditionable) GetConditionOfType(ttype string) (metav1.Condition, bool) {
   // ...
}
```

wdyt? maybe that's a different way of getting towards the same evaluation but with less of that code in this scope

Comment (idoru) [reply pkg/realizer/healthcheck/healthcheck.go:54]: Yeah, I was on the fence, but I didn't want to choose between adding a dependency on `mapstructure` or marshalling the map just to unmarshal it back into the struct. I improve the re-use of the private `conditionList` in realizer in the follow-up PR for owner `ResourcesHealthy` and I would still like to remove the json path query here.

File: pkg/realizer/healthcheck/healthcheck.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/vmware-tanzu/cartographer/pkg/utils"
+)
+
+func DetermineHealthCondition(rule *v1alpha1.HealthRule, realizedResource *v1alpha1.RealizedResource, stampedObject *unstructured.Unstructured) metav1.Condition {
+	if rule == nil {
+		if realizedResource == nil {
+			return conditions.NoResourceResourcesHealthyCondition()
+		} else if len(realizedResource.Outputs) > 0 {
+			return conditions.OutputAvailableResourcesHealthyCondition()
+		} else if realizedResource.TemplateRef != nil && realizedResource.TemplateRef.Kind == "ClusterTemplate" && realizedResource.TemplateRef.APIVersion == "carto.run/v1alpha1" {
```
Comment (cirocosta) [inline pkg/realizer/healthcheck/healthcheck.go]: uuuuhh it seems a little fragile to have these strings right here 😬 

wdyt of leveraging v1alpha1.SchemaGroupVersion

https://github.com/vmware-tanzu/cartographer/blob/693036beb08f237ad45e3715c92a61a7e718cd81/pkg/apis/v1alpha1/groupversion_info.go#L23-L26

for APIVersion? I think the `.Identifier()` or `.String()` method of that struct would do it

given that various `&&`,  might be useful to turn the predicate into a small function that returns a bool, wdyt?

**Review by cirocosta — APPROVED** [review]
lgtm! left just a nit comment, but ack'ed! feel free to merge as is or tackle the comment - on you 😁   thanks!

**Pull request merged**